HistogramData rollout: algorithms 76 to 81

raquelalvarezbanos opened the pull request:
Description of work.

**To test:**

<!-- Instructions for testing. -->

Fixes #17710

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

SimonHeybrock commented:
@raquelalvarezbanos Would you like me to have a look at your changes or are you still going to make substantial modifications?

raquelalvarezbanos commented:
Thanks @SimonHeybrock. I don't plan to make substantial modifications. I only need to figure out what is wrong with `RefinePowderInstrumentParameters`'s performance test. The error is:

/Users/builder/Jenkins/workspace/pull_requests-osx/Framework/CurveFitting/test/Algorithms/RefinePowderInstrumentParameters3Test.h:485: createSuite() already declared

I remember seeing that locally but in the end it worked. This is the only thing I wanted to fix, but it is OK if you want to start reviewing.

Also, I am seeing strange numbers in terms of performance: none of the algorithms for which I wrote performance tests showed significant improvement (which I guess can be normal). However, for `NormaliseByPeakArea` I am seeing a decrease in performance: `NormaliseByPeakAreaTestPerformance::test_sum_false()` took ~1.31s to run on my machine (Windows, release mode) before the changes, and ~3.86s after the changes. I guess I must be doing something wrong. `NormaliseByPeakAreaTestPerformance::test_sum_true()` also showed a decrease in performance, but only slightly.

SimonHeybrock commented on Framework/CurveFitting/test/Algorithms/RefinePowderInstrumentParameters3Test.h:
```diff
@@ -204,289 +470,75 @@ class RefinePowderInstrumentParameters3Test : public CxxTest::TestSuite {
-      throw runtime_error("generateInstrumentProfile supports bank 1 only.");
-    }
+class RefinePowderInstrumentParameters3TestPerformance
+    : public CxxTest::TestSuite {
```
@raquelalvarezbanos Someone else had the same problem recently: cxxtest beaks if the class declaration spans 2 lines (but clang-format forces it to two lines). In the end we used this trick that saves a few characters, see if that fits:

```cpp
using CxxTest::TestSuite;
class RefinePowderInstrumentParameters3TestPerformance : public TestSuite {
```

raquelalvarezbanos replied:
Yes, it fits, thanks!

SimonHeybrock commented on Framework/CurveFitting/inc/MantidCurveFitting/Algorithms/LeBailFit.h:
```diff
@@ -267,9 +272,9 @@ class DLLExport LeBailFit : public API::Algorithm {
-                     std::vector<double> &background);
+  // std::vector<double> fitBackground(size_t wsindex,
+  //                                  API::FunctionDomain1DVector domain,
+  //                                  API::FunctionValues values);
```
Remove?

raquelalvarezbanos replied:
I was tempted to remove it, but I didn't in the end because of [this](https://github.com/mantidproject/mantid/blob/17710_HistogramData_Rollout_Algorithms_76-81_new/Framework/CurveFitting/src/Algorithms/LeBailFit.cpp#L2413) comment. I don't know if anyone is planning to do work on this... Methods `LeBailFit::smoothBackgroundExponential` and `LeBailFit::smoothBackgroundAnalytical` also contain dead code which I didn't remove.

SimonHeybrock replied:
Hmm, that was added at least two years ago, I'd just remove these things.

raquelalvarezbanos replied:
OK, I'll remove dead code and code that is never called.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/FitPowderDiffPeaks.cpp:
```diff
@@ -3339,7 +3315,7 @@ bool observePeakParameters(Workspace2D_sptr dataws, size_t wsindex,
  * @return index of the maximum value
  */
-size_t findMaxValue(const MantidVec Y) {
+size_t findMaxValue(const std::vector<double> &Y) {
```
Is this used for anything but `HistogramY`? If not you could change the argument type and avoid the `rawData()` call above.

SimonHeybrock replied:
On second thought, could `std::max_element` be used here (with `std::distance`)?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -446,15 +447,15 @@ void LeBailFit::processInputBackground() {
-  const MantidVec &vecX = m_dataWS->readX(m_wsIndex);
-  MantidVec &vecY = m_outputWS->dataY(CALDATAINDEX);
+  const auto &vecX = m_dataWS->x(m_wsIndex).rawData();
+  std::vector<double> vecY(m_outputWS->y(CALDATAINDEX).size(), 0);
```
Why is this being zero-initialized now, instead of copied? Is it just overwritten below?

SimonHeybrock replied:
Having read a bit more below: You had to get into some mess to get the new data types into `calculateDiffractionPattern`. Have you considered changing the parameter types of the method to `HistogramX` and `HistogramY` instead of `std::vector<double>`?  I think that would make things much easier at the call sites.

raquelalvarezbanos replied:
Yes, that was exactly the problem. As you say, I guess updating the method to take `HistogramX` (and `Y`, etc) is the best solution. I'll change that.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -471,31 +472,34 @@ void LeBailFit::execPatternCalculation() {
   // Calculate background
-  MantidVec &vec_bkgd = m_outputWS->dataY(INPUTBKGDINDEX);
+  std::vector<double> vec_bkgd(m_outputWS->y(INPUTBKGDINDEX).size(), 0);
   m_lebailFunction->function(vec_bkgd, vecX, false, true);
```
Consider adapting the function signature of `LeBailFunction::function`. It does not seem to be used outside this algorithm so it would be a local change and would make your life easier here, I think.

raquelalvarezbanos replied:
Yes, my mistake, sorry.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -471,31 +472,34 @@ void LeBailFit::execPatternCalculation() {
-        m_outputWS->readY(OBSDATAINDEX)[i] -
-        m_outputWS->readY(INPUTBKGDINDEX)[i];
+    m_outputWS->mutableY(INPUTPUREPEAKINDEX)[i] =
+        m_outputWS->y(OBSDATAINDEX)[i] - m_outputWS->y(INPUTBKGDINDEX)[i];
```
I think you can just write `m_outputWS->mutableY(INPUTPUREPEAKINDEX) = m_outputWS->y(OBSDATAINDEX) - m_outputWS->y(INPUTBKGDINDEX)`.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -471,31 +472,34 @@ void LeBailFit::execPatternCalculation() {
-        m_outputWS->readY(OBSDATAINDEX)[i] - m_outputWS->readY(CALDATAINDEX)[i];
+    m_outputWS->mutableY(DATADIFFINDEX)[i] =
+        m_outputWS->y(OBSDATAINDEX)[i] - m_outputWS->y(CALDATAINDEX)[i];
   }
```
Same here.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -471,31 +472,34 @@ void LeBailFit::execPatternCalculation() {
     for (size_t ipk = 0; ipk < m_lebailFunction->getNumberOfPeaks(); ++ipk) {
-      MantidVec &vecTemp = m_outputWS->dataY(9 + ipk);
+      std::vector<double> vecTemp(m_outputWS->y(9 + ipk).size(), 0);
       m_lebailFunction->calPeak(ipk, vecTemp, vecX);
```
Adapt the function signature?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -551,15 +555,15 @@ void LeBailFit::execRefineBackground() {
   vector<double> backgroundvalues(numpts);
```
```cpp
vector<double> backgroundvalues(values);
```
?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -579,14 +583,14 @@ void LeBailFit::execRefineBackground() {
     m_backgroundFunction->function(domain, values);
     for (size_t i = 0; i < numpts; ++i) {
       backgroundvalues[i] = values[i];
```
Why not `backgroundvalues = values;` outside the loop?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -632,29 +636,29 @@ void LeBailFit::execRefineBackground() {
   m_backgroundFunction->function(domain, values);
   for (size_t i = 0; i < numpts; ++i) {
     backgroundvalues[i] = values[i];
```
Assign vectors?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -632,29 +636,29 @@ void LeBailFit::execRefineBackground() {
-    m_outputWS->dataY(1)[i] = valueVec[i] + backgroundvalues[i];
-    m_outputWS->dataY(2)[i] = vecY[i] - (valueVec[i] + backgroundvalues[i]);
+    m_outputWS->mutableY(1)[i] = valueVec[i] + backgroundvalues[i];
+    m_outputWS->mutableY(2)[i] = vecY[i] - (valueVec[i] + backgroundvalues[i]);
```
In general, calling `dataY` or `mutableY` in a loop over bins should be avoided, since it has non-negligible cost due to the copy-on-write mechanism. Here it is probably fine since it is done only for 2 histograms, but if you have a big workspace with many histograms you can see the difference.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -632,29 +636,29 @@ void LeBailFit::execRefineBackground() {
-    m_outputWS->dataY(CALPUREPEAKINDEX)[i] = valueVec[i];
+    m_outputWS->mutableY(CALBKGDINDEX)[i] = backgroundvalues[i];
+    m_outputWS->mutableY(CALPUREPEAKINDEX)[i] = valueVec[i];
   }
```
Wouldn't `=` without loop work here?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -1478,14 +1479,16 @@ void LeBailFit::execRandomWalkMinimizer(size_t maxcycles,
   // Process background to make a pure peak spectrum in output workspace
-  MantidVec &vecBkgd = m_outputWS->dataY(INPUTBKGDINDEX);
+  std::vector<double> vecBkgd(m_outputWS->y(INPUTBKGDINDEX).size(), 0);
   m_lebailFunction->function(vecBkgd, vecX, false, true);
```
See above, can avoid some copies if you change the function signature.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -1508,10 +1511,10 @@ void LeBailFit::execRandomWalkMinimizer(size_t maxcycles,
+  auto &vecCalBkgd = m_outputWS->mutableY(CALBKGDINDEX);
   for (size_t i = 0; i < numpts; ++i) {
     // Calculated (refined) data
     vecCalY[i] = vecCalPurePeaks[i] + vecBkgd[i];
```
There are `operator=` and `operator+` etc. for the types in `HistogramData`, so you can avoid the loop.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -1961,29 +1964,29 @@ void LeBailFit::addParameterToMCMinimize(vector<string> &parnamesforMC,
+* @return :: boolean value.  whether all the peaks' parameters are physical.
+*/
+bool LeBailFit::calculateDiffractionPattern(const std::vector<double> &vecX,
+                                            const std::vector<double> &vecY,
```
Have you considered changing these types to `HistgramX`, etc.?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/NormaliseByPeakArea.cpp:
```diff
@@ -283,29 +275,31 @@ double NormaliseByPeakArea::fitToMassPeak(const MatrixWorkspace_sptr &yspace,
+  std::vector<double> outY(m_normalisedWS->y(index).size());
   std::transform(inY.begin(), inY.end(), outY.begin(),
                  std::bind2nd(std::divides<double>(), area));
+  m_normalisedWS->mutableY(index) = outY;
```
Why are you creating a temporary vector?

You can either do the transform on `HistogramY`, or use `operator/=`:

```cpp
m_normalisedWS->mutableY(index) /= area;
```

The latter might be slightly slower than doing the transform directly.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/NormaliseByPeakArea.cpp:
```diff
@@ -283,29 +275,31 @@ double NormaliseByPeakArea::fitToMassPeak(const MatrixWorkspace_sptr &yspace,
+  std::vector<double> outE(m_normalisedWS->e(index).size());
   std::transform(inE.begin(), inE.end(), outE.begin(),
                  std::bind2nd(std::divides<double>(), area));
+  m_normalisedWS->mutableE(index) = outE;
```
Same here. But actually you could just use `Histogram &operator/=(Histogram &histogram, const double factor)` (include `MantidHistogramData/HistogramMath.h`) and get rid of the entire function, I think?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/NormaliseByPeakArea.cpp:
```diff
@@ -283,29 +275,31 @@ double NormaliseByPeakArea::fitToMassPeak(const MatrixWorkspace_sptr &yspace,
+*/
 void NormaliseByPeakArea::saveToOutput(const API::MatrixWorkspace_sptr &accumWS,
                                        const std::vector<double> &yValues,
                                        const std::vector<double> &eValues,
```
Consider changing the function parameter types, that will make things easier at the call sites of this function.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/NormaliseByPeakArea.cpp:
```diff
@@ -314,10 +308,10 @@ void NormaliseByPeakArea::saveToOutput(const API::MatrixWorkspace_sptr &accumWS,
-    auto &accumY = accumWS->dataY(0);
-    auto &accumE = accumWS->dataE(0);
-    const auto accumYCopy = accumWS->readY(0);
-    const auto accumECopy = accumWS->readE(0);
```
Why is this making a copy of the vectors? Do you see a reason that I am overlooking?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/NormaliseByPeakArea.cpp:
```diff
@@ -354,11 +348,11 @@ void NormaliseByPeakArea::symmetriseYSpace() {
-    auto &ySymOut = m_symmetrisedWS->dataY(i);
-    auto &eSymOut = m_symmetrisedWS->dataE(i);
-    const auto yIn = m_yspaceWS->readY(i); // copy
-    const auto eIn = m_yspaceWS->readE(i); // copy
```
Again, why?

raquelalvarezbanos replied:
No idea, I'll remove that.

SimonHeybrock commented on Framework/CurveFitting/test/Algorithms/FitPowderDiffPeaksTest.h:
```diff
@@ -107,8 +436,18 @@ class FitPowderDiffPeaksTest : public CxxTest::TestSuite {
+    TS_ASSERT_EQUALS(peakdataws->x(0).rawData(), peakdataws->x(1).rawData());
+    TS_ASSERT_EQUALS(peakdataws->x(0).rawData(), peakdataws->x(2).rawData());
+    TS_ASSERT_EQUALS(peakdataws->x(0).rawData(), peakdataws->x(3).rawData());
+    TS_ASSERT_EQUALS(peakdataws->x(0).rawData(), peakdataws->x(4).rawData());
```
There is an `operator==` for these types available in `#include "MantidTestHelpers/HistogramDataTestHelper.h"` (it is not included by default since we do not want to do exact vector comparison in general code). But using `rawData` is also fine :+1:.

raquelalvarezbanos replied:
Thanks, I didn't know, but if it exists, I'll use it :)

SimonHeybrock commented on Framework/CurveFitting/test/Algorithms/FitPowderDiffPeaksTest.h:
```diff
@@ -117,355 +456,90 @@ class FitPowderDiffPeaksTest : public CxxTest::TestSuite {
-    // 3. Output information
-    std::cout << "Created Table Workspace with " << hklws->rowCount()
-              << " entries of peaks.\n";
+  void setUp() {
```
Use the `override` keyword for `setUp` and `tearDown` please. We will move the Mantid builds to Ubuntu 16.04 soon and the newer gcc creates warnings otherwise, which makes the builds fail.

SimonHeybrock replied:
Still missing an `override` here (unless the github diff is messed up).

SimonHeybrock commented on Framework/CurveFitting/test/Algorithms/FitPowderDiffPeaksTest.h:
```diff
@@ -117,355 +456,90 @@ class FitPowderDiffPeaksTest : public CxxTest::TestSuite {
-    newrow22 << "Zerot" << 499.99;
-
-    return;
+  void tearDown() {
```
`override`

raquelalvarezbanos replied:
:+1:

SimonHeybrock replied:
`override`

SimonHeybrock commented on Framework/CurveFitting/test/Algorithms/LeBailFitTest.h:
```diff
@@ -30,6 +21,642 @@ using namespace std;
+    dataws->dataX(0)[i] = vecX[i];
+    dataws->dataY(0)[i] = vecY[i];
+    dataws->dataE(0)[i] = vecE[i];
+  }
```
You can do this without temporary arrays:

```cpp
dataws->setHistogram(0, BinEdges{ /* initializer_list */ }, Counts {...}, CountStandardDeviations{...});
```

raquelalvarezbanos replied:
I am trying to do the following:

```
BinEdges vecX = { 70931.750000, 70943.609000, ...};
Counts vecY = {0.000000,    0.000000,    0.695623, ...};
CountStandardDeviations vecE = {1.000000,  1.000000,  1.000000, ...};

auto dataws = boost::dynamic_pointer_cast<API::MatrixWorkspace>(
      API::WorkspaceFactory::Instance().create("Workspace2D", 1, 127, 127));

dataws->setHistogram(0, vecX, vecY, vecE);
```

but I get an exception in `Histogram::checkSize()` [here](https://github.com/mantidproject/mantid/blob/master/Framework/HistogramData/inc/MantidHistogramData/Histogram.h#L399-L400). At that point `target` is 126 but all my vectors `vecX`, `vecY`, `vecE`, were 127, I guess it is because as `vecX` is `BinEdges` it is expected to have an extra element?

SimonHeybrock replied:
@raquelalvarezbanos Sorry, my bad, you need `Points` instead of `BinEdges` (as I should have seen from the workspace creation which uses `size` for both X and Y length).

raquelalvarezbanos replied:
:+1:

SimonHeybrock commented on Framework/CurveFitting/test/Algorithms/LeBailFitTest.h:
```diff
@@ -30,6 +21,642 @@ using namespace std;
+    dataws->dataX(0)[i] = vecx[i];
+    dataws->dataY(0)[i] = vecy[i];
+    dataws->dataE(0)[i] = vece[i];
+  }
```
Could use `setHistogram`.

SimonHeybrock commented on Framework/CurveFitting/test/Algorithms/LeBailFitTest.h:
```diff
@@ -30,6 +21,642 @@ using namespace std;
+    double e = 1.0;
+    if (y > 1.0)
+      e = sqrt(y);
+    vecE[i] = e;
```
`vecE[i] = std::max(1.0, sqrt(vecY[i]));`?

SimonHeybrock commented on Framework/CurveFitting/test/Algorithms/RefinePowderInstrumentParameters3Test.h:
```diff
@@ -21,6 +21,289 @@ using namespace Mantid::API;
+        if (y > 1.0)
+          e = std::sqrt(y);
+        else
+          e = 1.0;
```
`std::max`?

raquelalvarezbanos commented:
Thanks @SimonHeybrock I'll start adding your suggested changes.

raquelalvarezbanos commented:
@SimonHeybrock : thanks for the thorough testing, I think have addressed all your suggestions, but please let me know if I missed anything. I'll measure again performance.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -1478,14 +1464,16 @@ void LeBailFit::execRandomWalkMinimizer(size_t maxcycles,
-  m_lebailFunction->function(vecBkgd, vecX, false, true);
-  MantidVec &dataPurePeak = m_outputWS->dataY(INPUTPUREPEAKINDEX);
+  std::vector<double> vecBkgd =
+      m_lebailFunction->function(vecX.rawData(), false, true);
```
Remove `rawData`?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFit.cpp:
```diff
@@ -1478,14 +1464,16 @@ void LeBailFit::execRandomWalkMinimizer(size_t maxcycles,
+  m_outputWS->mutableY(INPUTBKGDINDEX) = vecBkgd;
+  std::vector<double> dataPurePeak(m_outputWS->y(INPUTPUREPEAKINDEX).size(), 0);
   transform(vecInY.begin(), vecInY.end(), vecBkgd.begin(), dataPurePeak.begin(),
             ::minus<double>());
```
Why not store the result of `m_lebailFunction->function` in a `HistogramY` (or directly return one?)? THen you can simply use `operator-` or `-=` here.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFunction.cpp:
```diff
@@ -78,33 +79,32 @@ LeBailFunction::LeBailFunction(std::string peaktype) {
+* otherwise, assume zero background
+* @return :: output vector
+*/
+std::vector<double>
```
As mentioned above, consider returning `HistogramY`, unless that does not work somewhere?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/LeBailFunction.cpp:
```diff
@@ -135,12 +135,16 @@ void LeBailFunction::function(std::vector<double> &out,
-void LeBailFunction::calPeak(size_t ipk, std::vector<double> &out,
-                             const std::vector<double> &xvalues) const {
+*/
+std::vector<double> LeBailFunction::calPeak(size_t ipk,
```
Have you considered returning `HistogramY`?

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/NormaliseByPeakArea.cpp:
```diff
@@ -283,45 +275,38 @@ double NormaliseByPeakArea::fitToMassPeak(const MatrixWorkspace_sptr &yspace,
-                                       const std::vector<double> &yValues,
-                                       const std::vector<double> &eValues,
+                                       const HistogramData::HistogramY &yValues,
+                                       const HistogramData::HistogramE &eValues,
```
Look at how this function is called, and how Y and E are used below --- at least sometimes data could be shared, so consider passing shared pointers here, i.e., the result of `sharedY` etc, and then use `setSharedY` below.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/NormaliseByPeakArea.cpp:
```diff
@@ -305,8 +307,8 @@ void NormaliseByPeakArea::saveToOutput(const API::MatrixWorkspace_sptr &accumWS,
-      double rhsYj = yValues[j];
-      double rhsEj = eValues[j];
+      double rhsYj = yValues->rawData()[j];
+      double rhsEj = eValues->rawData()[j];
```
Consider moving the call to `rawData` outside the loop. Just like the old `dataY()` methods (and as done for `accumWS->mutableY()` above), it is better to buffer the reference since the copy-on-write mechanism makes write access more expensive.

SimonHeybrock commented on Framework/CurveFitting/src/Algorithms/NormaliseByPeakArea.cpp:
```diff
@@ -303,12 +303,14 @@ void NormaliseByPeakArea::saveToOutput(
     auto &accumY = accumWS->mutableY(0);
     auto &accumE = accumWS->mutableE(0);
+    auto yValuesRaw = yValues->rawData();
+    auto eValuesRaw = eValues->rawData();
```
Reference?

SimonHeybrock commented:
👍 
:shipit: once Jenkins gives his ok.

raquelalvarezbanos commented:
Thanks again @SimonHeybrock for testing!

SimonHeybrock commented:
@raquelalvarezbanos No problem, you're welcome. Thanks for the work!

martyngigg commented:
@raquelalvarezbanos Sorry that this has been lingering around. It looks like the Ubuntu build has some warnings to take care of. Once those are taken care of I can merge it.

raquelalvarezbanos commented:
Thanks @martyngigg !
